python/ray/autoscaler/_private/kuberay/node_provider.py

```diff
@@ -0,0 +1,234 @@
+# the Ray autoscaler they are called "available node types".
+
+## Design:
+
+# Each modification the autoscaler wants to make is posted to the API server goal state
+# (e.g. if the autoscaler wants to scale up, it increases the number of
+# replicas of the worker group it wants to scale, if it wants to scale down
+# it decreases the number of replicas and adds the exact pods that should be
+# terminated to the scaleStrategy). In order to guarantee consistency, the NodeProvider
+# then waits until Kuberay's reconciliation loop creates the pods and then returns control
+# back to the autoscaler. The waiting period is typically small, on the order
+# of a few seconds. We make sure that only one such modification is in process
+# by serializing all modification operations with a lock in the NodeProvider.
+
+
+# Note: Log handlers set up in autoscaling monitor entrypoint.
+logger = logging.getLogger(__name__)
+
+provider_exists = False
+
+
+def to_label_selector(tags):
+    label_selector = ""
+    for k, v in tags.items():
+        if label_selector != "":
+            label_selector += ","
+        label_selector += "{}={}".format(k, v)
+    return label_selector
+
+def status_tag(pod: Dict[str, Any]):
+    if "containerStatuses" not in pod["status"]:
+        return "pending"
+    
+    state = pod["status"]["containerStatuses"][0]["state"]
+
+    logger.info("status_tag: state = {}".format(state))
+
+    if "pending" in state:
+        return "pending"
+    if "running" in state:
+        return "up-to-date"
+    if "waiting" in state:
+        return "waiting"
+    if "terminated" in state:
+        return "update-failed"
+    raise ValueError("Unexpected container state.")
+
+def make_node_tags(labels: Dict[str, str], status_tag: str):
+    "Convert Kuberay labels to Ray autoscaler tags."
+    tags = {"ray-node-status": status_tag}
+
+    if labels["ray.io/node-type"] == "head":
+        tags["ray-node-type"] = "head"
+        tags["ray-user-node-type"] = "head-group"
+    else:
+        tags["ray-node-type"] = "worker"
+        tags["ray-user-node-type"] = labels["ray.io/group"]
+
+    return tags
+
+
+class KuberayNodeProvider(NodeProvider):  # type: ignore
+    def __init__(
+        self,
+        provider_config: Dict[str, Any],
+        cluster_name: str,
+        _allow_multiple: bool = False,
+    ):
+        logger.info("Creating ClientNodeProvider.")
+        self.namespace = provider_config["namespace"]
+        self.cluster_name = cluster_name
+        self._lock = threading.RLock()
+
+        with open("/var/run/secrets/kubernetes.io/serviceaccount/token") as secret:
+            token = secret.read()
+
+        self.headers = {
+            "Authorization": "Bearer " + token,
+        }
+        self.verify = "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+
+        # Disallow multiple node providers, unless explicitly allowed for testing.
+        global provider_exists
+        if not _allow_multiple:
```

Comment:
What's the purpose of multiple provider? that means in future people want a centralize autoscaler with mutli-node-provider support?  Is that the direction?

Reply:
We are using that for our unit tests to test/mock multiple node providers in one process as far as I know. We are not planning to support multiple providers in production, at least not any time soon :)

Reply:
Yep -- provider should be a singleton per process, making an exception for unit tests.